### DISCUSSION TITLE
Add generic constraint for non-abstract (and abstract)

### DISCUSSION THREAD

**Discussion by zorgoz** [dotnet/csharplang#8391] [Language Ideas]
Before `required` property modifier was added to the language, the constraint `new()` was practically equivalent to a _non-abstract_ constraint. But currently there is no combination constraints equivalent to restriction to non-abstract types that may have required properties. 

Suggested keywords: `concrete` and `abstract`. 

-  `abstract` and `new()` are mutually exclusive
- the new constraints cannot be combined with `struct`

**Comment by huoyaoyuan** [comment]
Generic constraints provided by the language enables additional operations under the constraint. Constraints that doesn't enable operation but purely limiting usages should be provided by analyzers.

**Reply by Joe4evr** [reply]
[I even made one like that.](https://github.com/Joe4evr/JiiLib/blob/master/src/JiiLib.Constraints/Analyzers/NonAbstractOnly/NonAbstractConstraintAnalyzer.cs) I just haven't had the time to improve it more.

**Comment by zorgoz** [comment]
That's nice. Thanks for the hint.

Joe4evr ***@***.***> ezt írta (időpont: 2024. aug. 27., K,
11:14):

**Comment by TahirAhmadov** [comment]
You can also do a static ctor to check it at runtime:
```cs
class Foo<T>
{
  static Foo()
  {
    if(typeof(T).IsAbstract) throw new Exception("Type T cannot be abstract");
  }
}
```

**Comment by zorgoz** [comment]
I know that, but for my purposes a design-time approach is way better. I
suppose I will go with analyzers as I have way more things to check.

Mr. TA ***@***.***> ezt írta (időpont: 2024. aug. 27., K,
19:29):
